### PR TITLE
Created ObjectSquareMatrix

### PR DESCRIPTION
**Pull request by Beakerboy**
With this class, matrix addition, subtraction, and determinants can be performed on objects.

### CODE REVIEW

**Comment by Beakerboy** [comment]
I have not implemented Matrix multiplication yet. The parent allows a matrix to be multiplied by a Vector. if you wish to maintain that, we will have to perform the Vector::dotProduct() algorithm within the ObjectMatrix::multiply() method to avoid creating an ObjectVector class. Thankfully, it looks like the Vector class will not complain if it is constructed with objects. Many of the native method would fail though.

File: src/LinearAlgebra/ObjectSquarematrix.php [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+     *
+     * @throws BadDataExpcetion if any rows have a different column count
+     * @throws IncorrectTypeException if all elements are not the same class
+     * @throws IncorrectTypeException if The class does not implement the ObjectArithmetic interface 
+     */
+    public function __construct(array $A)
+    {
+        parent::__construct($A);
+
+        if (in_array("MathPHP\Number\ObjectArithmetic", class_implements($A[0][0]))) {
```
Comment (markrogoyski) [inline src/LinearAlgebra/ObjectSquarematrix.php]: You can use ```instanceof``` to check if an object implements an interface. It seems cleaner than matching a string against an array.

```php
if ($A[0][0] instanceof MathPHP\Number\ObjectArithmetic) {
   ...
}

File: src/LinearAlgebra/ObjectSquarematrix.php [hunk 2]
```diff
@@ -0,0 +1,172 @@
+<?php
```
Comment (markrogoyski) [inline src/LinearAlgebra/ObjectSquarematrix.php]: PSR autoloading and style guides say the file name and class name have to match. matrix is not CamelCased in the file name.
ObjectSquarematrix -> ObjectSquareMatrix

File: tests/LinearAlgebra/ObjectSquareMatrixTest.php [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        $A = MatrixFactory::create([[$object]]);
+    }
+
+    public function testMatrixAddException()
+    {
+        $polynomial = new Polynomial([1, 4, 7]);
+        $complex = new Complex(1, 4);
+        $A = MatrixFactory::create([[$polynomial]]);
+        $B = MatrixFactory::create([[$complex]]);
+        $this->expectException(Exception\IncorrectTypeException::class);
```
Comment (markrogoyski) [inline tests/LinearAlgebra/ObjectSquareMatrixTest.php]: Thanks for using the more modern format.

File: src/LinearAlgebra/ObjectSquarematrix.php [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    {
+        if ($B->getM() !== $this->m || $B->getN() !== $this->n) {
+            throw new \MathPHP\Exception\MatrixException('Matrices are different sizes');
+        }
+        if ($B->getObjectType() !== $this->object_type) {
+            throw new \MathPHP\Exception\IncorrectTypeException('Matrices must contain the same object types');
+        }
+    }
+
+    /*
```
Comment (markrogoyski) [inline src/LinearAlgebra/ObjectSquarematrix.php]: Minor thing: docBlock starts with ```/**```
If you use an IDE or SublimeText or something with syntax highlighting, usually the @tags will not syntax highlight properly without the second * so it is a visual clue that something is not right. Github does the same.

**Comment by markrogoyski** [comment]
Thanks for the pull request. Looks good. Nice tests. Just a couple of minor improvements to make.

Btw, we talked about removing the SquareMatrix. I'm fine with that. Doesn't need to be done now in this pull request. I imagine a lot of tests will break. If it is still something you think we should do let me know and I'll put it on my backlog of things to do.

File: src/LinearAlgebra/ObjectSquarematrix.php (hunk 1 above)
Comment (Beakerboy) [inline src/LinearAlgebra/ObjectSquarematrix.php]: Good call, I didn't know this worked on interfaces

**Comment by Beakerboy** [comment]
Yeah, I think reorganizing the inheritance of the Matrix classes will be helpful in the long term. We can use traits to manage multiple-inheritance, and just have everything inherit the main Matrix class. If you have a "Backlog", can you use the GitHub Projects manager to list them?

**Comment by markrogoyski** [comment]
I'll check out the Projects manager. Thanks.

**Comment by markrogoyski** [comment]
I created a development backlog project: https://github.com/markrogoyski/math-php/projects/1

**Comment by Beakerboy** [comment]
I see you have "covariant matrix" on the list. The Matrix class already has a method to calculate the covariance matrix.

**Pull request merged**